packages/core/src/types.ts

```diff
@@ -75,6 +81,365 @@ export interface PositionWithOffset extends Position {
+
+	/**
+	 * Removes the specified node or token.
+	 * @param nodeOrToken The node or token to remove.
+	 * @returns The edit to remove the node or token.
+	 */
+	remove(nodeOrToken: object): RuleTextEdit;
+
+	/**
+	 * Removes the specified range.
+	 * @param range The range to remove.
+	 * @returns The edit to remove the range.
+	 */
+	removeRange(range: SourceRange): RuleTextEdit;
+
+	/**
+	 * Replaces the specified node or token with the given text.
+	 * @param nodeOrToken The node or token to replace.
+	 * @param text The text to replace the node or token with.
+	 * @returns The edit to replace the node or token.
+	 */
+	replaceText(nodeOrToken: object, text: string): RuleTextEdit;
+
+	/**
+	 * Replaces the specified range with the given text.
+	 * @param range The range to replace.
+	 * @param text The text to replace the range with.
+	 * @returns The edit to replace the range.
+	 */
+	replaceTextRange(range: SourceRange, text: string): RuleTextEdit;
+}
+
+/**
+ * Represents a fix for a rule violation implemented as a text edit.
+ */
+export interface RuleTextEdit {
+	/**
+	 * The range to replace.
+	 */
+	range: SourceRange;
+
+	/**
+	 * The text to insert.
+	 */
+	text: string;
+}
+
+// #endregion
+
+interface ViolationReportBase {
+	/**
+	 * The type of node that the violation is for.
+	 * @deprecated May be removed in the future.
+	 */
+	nodeType?: string | undefined;
+
+	/**
+	 * The data to insert into the message.
+	 */
+	data?: Record<string, string> | undefined;
+
+	/**
+	 * The fix to be applied for the violation.
+	 * @param fixer The text editor to apply the fix.
+	 * @returns The fix(es) for the violation.
+	 */
+	fix?(
+		fixer: RuleTextEditor,
+	): RuleTextEdit | RuleTextEdit[] | IterableIterator<RuleTextEdit> | null;
+
+	/**
+	 * An array of suggested fixes for the problem. These fixes may change the
+	 * behavior of the code, so they are not applied automatically.
+	 */
+	suggest?: Array<SuggestedEdit>;
+}
+
+type ViolationMessage = { message: string } | { messageId: string };
+type ViolationLocation = { loc: SourceLocation } | { node: object };
+
+export type ViolationReport = ViolationReportBase &
+	ViolationMessage &
+	ViolationLocation;
+
+// #region Suggestions
+
+interface SuggestedEditBase {
+	/**
+	 * The data to insert into the message.
+	 */
+	data?: Record<string, string> | undefined;
+
+	/**
+	 * The fix to be applied for the suggestion.
+	 * @param fixer The text editor to apply the fix.
+	 * @returns The fix for the suggestion.
+	 */
+	fix?(
+		fixer: RuleTextEditor,
+	): RuleTextEdit | RuleTextEdit[] | IterableIterator<RuleTextEdit> | null;
+}
+
+type SuggestionMessage = { desc: string } | { messageId: string };
+
+/**
+ * A suggested edit for a rule violation.
+ */
+export type SuggestedEdit = SuggestedEditBase & SuggestionMessage;
+
+// #endregion
+
+/**
+ * The definition of an ESLint rule.
+ */
+export interface RuleDefinition<T extends RuleVisitor = RuleVisitor> {
```

Comment:
Using a type parameter here allows language plugins to insert their own `RuleVisitor` interface for better type checking. I imagine we can turn the one in `@types/eslint` into a `RuleVisitor` so those definitions will still work.